include/seqan/bam_io/bam_index_bai.h

```diff
@@ -154,6 +155,164 @@ class BamIndex<Bai>
+
+    list.clear();
+
+    unsigned k;
+    list.push_back(0);
+    for (k =    1 + (beg>>26); k <=    1 + (end>>26); ++k) list.push_back(k);
+    for (k =    9 + (beg>>23); k <=    9 + (end>>23); ++k) list.push_back(k);
+    for (k =   73 + (beg>>20); k <=   73 + (end>>20); ++k) list.push_back(k);
+    for (k =  585 + (beg>>17); k <=  585 + (end>>17); ++k) list.push_back(k);
+    for (k = 4681 + (beg>>14); k <= 4681 + (end>>14); ++k) list.push_back(k);
+}
+
+inline bool
+_getMinFileOffset(uint64_t & minFileOffset,
+                  BamIndex<Bai> const & bamIndex,
+                  int32_t const refId,
+                  uint32_t const position)
+{
+    uint32_t windowIdx = position >> 14; // Linear index consists of 16kb windows.
+
+    if (windowIdx >= length(bamIndex._linearIndices[refId]))
+        return false;
+
+    minFileOffset = bamIndex._linearIndices[refId][windowIdx];
+
+    return true;
+}
+
+template <typename TContainer, typename TSpec>
+inline
+SEQAN_FUNC_ENABLE_IF(And<IsSameType<typename Value<TContainer>::Type, BamAlignmentRecord>,
+                     Not<IsSameType<typename Value<TContainer>::Type, TContainer>>>, void)
+viewRecords(TContainer & resultContainer,
+            FormattedFile<Bam, Input, TSpec> & bamFile,
+            BamIndex<Bai> const & bamIndex,
+            int32_t const refId,
+            uint32_t regionStart,
+            uint32_t regionEnd)
+{
+    typedef typename std::map<uint32_t, BaiBamIndexBinData_>::const_iterator TMapIter;
+    typedef String<Pair<uint64_t, uint64_t>> TChunk;
+    typedef Iterator<TChunk const, Rooted>::Type TChunkIter;
+
+    // Sanity checks
+    // -------------------------------------------------------------------------
+    if (!isEqual(format(bamFile), Bam()))
+        return;
+
+    if (refId < 0 ||
+        static_cast<decltype(length(contigNames(context(bamFile))))>(refId) > length(contigNames(context(bamFile))))
+        return;  // Cannot seek to invalid reference.
+
+    if (static_cast<decltype(length(bamIndex._binIndices))>(refId) >= length(bamIndex._binIndices))
+        return;  // Cannot seek to invalid reference.
+
+    // correct end positions that exceed the reference length.
+    if (regionEnd > static_cast<uint32_t>(contigLengths(context(bamFile))[refId]))
+        regionEnd = static_cast<uint32_t>(contigLengths(context(bamFile))[refId]);
+
+    if (regionStart > regionEnd)
+        return; // invalid region
+
+    // Adjust region
+    // -------------------------------------------------------------------------
+    // We want to mimic samtools view. Since SAM/BAM files are 1-based we expect
+    // the given region to be also 1-based, to be consistent with samtools.
+    // Since SeqAn works 0-based we need to subtract 1.
+    SEQAN_ASSERT_GEQ_MSG(regionStart, 1u, "Expect 1-based region interval which excludes 0 as a start pos.");
+
+    if (regionStart != 0) // avoid overflow in case of wrong usage of user
```

Comment:
I would assert this as well. Otherwise [0,1] would be a valid interval by the user but might give you unexpected results.